File: backend/app/routers/vault.py
"""RAG Content Vault REST API."""
from __future__ import annotations

import uuid
from datetime import datetime, timezone

from fastapi import APIRouter, File, Form, HTTPException, Query, Request, UploadFile

from app.models.vault_schemas import (
    BatchDeleteRequest,
    BatchDeleteItem,
    BatchDeleteResult,
    BatchIngestRequest,
    BatchIngestResponse,
    BatchIngestSkippedItem,
    BatchStatusResponse,
    BatchUploadFileResult,
    BatchUploadResponse,
    ClearIndexResponse,
    CreateFileRequest,
    CreateFolderRequest,
    FolderListResponse,
    FolderRenamePreview,
    IngestPreviewItem,
    IngestPreviewRequest,
    IngestPreviewResponse,
    MigrateV16JobEntry,
    MigrateV16Response,
    PaginatedFilesResponse,
    RenameFolderRequest,
    ReindexResponse,
    SaveContentRequest,
    SyncReport,
    UploadResponse,
    VaultFile,
    VaultFolder,
    PAGE_SIZES,
)
from app.services import vault_db, vault_store
from app.services.ingest_estimate import preview_ingest_files
from app.services.job_queue import enqueue_vault_ingest
from app.services.vault_migration import any_ingest_locked, run_full_vault_migration
from app.services.vault_store import VaultStoreError
from app.services.vault_sync import sync_vault

router = APIRouter(prefix="/api/v1/rag/vault", tags=["vault"])

SYNC_STALE_SECONDS = 300


def _raise_store(exc: VaultStoreError) -> None:
    raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc


def _maybe_stale_sync() -> None:
    state = vault_db.get_sync_state()
    last = state.get("last_sync_at")
    if not last:
        sync_vault()
        return
    try:
        ts = datetime.fromisoformat(last.replace("Z", "+00:00"))
    except ValueError:
        sync_vault()
        return
    age = (datetime.now(timezone.utc) - ts).total_seconds()
    if age > SYNC_STALE_SECONDS:
        sync_vault()


def _enqueue_ingest_or_409(file_id: str, *, traceparent: str = "") -> str:
    try:
        return enqueue_vault_ingest(file_id, traceparent=traceparent)
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc


@router.get("/folders", response_model=FolderListResponse)
def get_folders() -> FolderListResponse:
    return FolderListResponse(folders=vault_store.list_folders())


@router.post("/folders", response_model=VaultFolder, status_code=201)
def post_folder(body: CreateFolderRequest) -> VaultFolder:
    try:
        return vault_store.create_folder(body.name)
    except VaultStoreError as exc:
        _raise_store(exc)
        raise  # pragma: no cover


@router.get("/folders/{folder_id}/rename-preview", response_model=FolderRenamePreview)
def get_folder_rename_preview(
    folder_id: str, name: str = Query(..., min_length=1, max_length=120)
) -> FolderRenamePreview:
    try:
        return vault_store.preview_folder_rename(folder_id, name)
    except VaultStoreError as exc:
        _raise_store(exc)
        raise  # pragma: no cover


@router.patch("/folders/{folder_id}", response_model=VaultFolder)
def patch_folder(folder_id: str, body: RenameFolderRequest) -> VaultFolder:
    try:
        return vault_store.rename_folder(folder_id, body.name)
    except VaultStoreError as exc:
        _raise_store(exc)
        raise  # pragma: no cover


@router.delete("/folders/{folder_id}", status_code=204)
def remove_folder(folder_id: str) -> None:
    try:
        vault_store.delete_folder(folder_id)
    except VaultStoreError as exc:
        _raise_store(exc)


@router.get("/files", response_model=PaginatedFilesResponse)
def get_files(
    folder_id: str | None = Query(default=None),
    keyword: str | None = Query(default=None),
    index_status: str | None = Query(default=None),
    search_content: bool = Query(default=False),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=10),
) -> PaginatedFilesResponse:
    if page_size not in PAGE_SIZES:
        raise HTTPException(
            status_code=422,
            detail=f"page_size must be one of {sorted(PAGE_SIZES)}",
        )
    _maybe_stale_sync()
    return vault_store.list_files(
        folder_id=folder_id,
        keyword=keyword,
        index_status=index_status,
        search_content=search_content,
        page=page,
        page_size=page_size,
    )


@router.post("/files", response_model=UploadResponse, status_code=201)
def post_file(
    body: CreateFileRequest,
    on_conflict: str = Query(default="replace"),
) -> UploadResponse:
    if on_conflict not in ("replace", "fail"):
        raise HTTPException(status_code=422, detail="on_conflict must be replace or fail")
    try:
        result = vault_store.create_text_file(
            folder_id=body.folder_id,
            filename=body.filename,
            content=body.content,
            source="created",
            on_conflict=on_conflict,  # type: ignore[arg-type]
        )
    except VaultStoreError as exc:
        _raise_store(exc)
        raise  # pragma: no cover
    return UploadResponse(file=result.file, ingest_job_id=None, replaced=result.replaced)


@router.post("/files/upload", response_model=UploadResponse, status_code=201)
async def upload_file(
    folder_id: str = Form(...),
    file: UploadFile = File(...),
    on_conflict: str = Query(default="replace"),
) -> UploadResponse:
    if on_conflict not in ("replace", "fail"):
        raise HTTPException(status_code=422, detail="on_conflict must be replace or fail")
    data = await file.read()
    filename = file.filename or "upload.md"
    try:
        result = vault_store.upload_file(
            folder_id=folder_id,
            filename=filename,
            data=data,
            on_conflict=on_conflict,  # type: ignore[arg-type]
        )
    except VaultStoreError as exc:
        _raise_store(exc)
        raise  # pragma: no cover
    except UnicodeDecodeError as exc:
        raise HTTPException(status_code=400, detail="File must be UTF-8 text") from exc
    return UploadResponse(
        file=result.file,
        ingest_job_id=None,
        replaced=result.replaced,
    )


@router.post("/files/batch-upload", response_model=BatchUploadResponse, status_code=201)
async def batch_upload(
    folder_id: str = Form(...),
    files: list[UploadFile] = File(...),
    on_conflict: str = Query(default="replace"),
) -> BatchUploadResponse:
    if on_conflict not in ("replace", "fail"):
        raise HTTPException(status_code=422, detail="on_conflict must be replace or fail")
    if not files:
        raise HTTPException(status_code=400, detail="No files provided")
    batch_id = str(uuid.uuid4())
    vault_db.insert_batch(batch_id=batch_id, total_files=len(files))
    results: list[BatchUploadFileResult] = []

    for upload in files:
        filename = upload.filename or "upload.md"
        data = await upload.read()
        try:
            result = vault_store.upload_file(
                folder_id=folder_id,
                filename=filename,
                data=data,
                on_conflict=on_conflict,  # type: ignore[arg-type]
            )
            vault_db.add_batch_file(batch_id, result.file.id, job_id=None)
            results.append(
                BatchUploadFileResult(
                    file_id=result.file.id,
                    filename=filename,
                    job_id=None,
                    status="uploaded" if not result.replaced else "replaced",
                )
            )
        except (VaultStoreError, UnicodeDecodeError, ValueError) as exc:
            results.append(
                BatchUploadFileResult(
                    file_id="",
                    filename=filename,
                    job_id=None,
                    status="failed",
                    error_message=str(exc),
                )
            )
            vault_db.refresh_batch_counts(batch_id)

    vault_db.refresh_batch_counts(batch_id)
    return BatchUploadResponse(batch_id=batch_id, files=results)


@router.post("/files/ingest-preview", response_model=IngestPreviewResponse)
def post_ingest_preview(body: IngestPreviewRequest) -> IngestPreviewResponse:
    raw = preview_ingest_files(body.file_ids)
    items = [IngestPreviewItem(**item) for item in raw]
    ingestible = [i for i in items if i.ingestible]
    return IngestPreviewResponse(
        items=items,
        total_estimated_tokens=sum(i.estimated_tokens for i in ingestible),
        file_count=len(ingestible),
    )


@router.post("/files/batch-ingest", response_model=BatchIngestResponse, status_code=202)
def post_batch_ingest(body: BatchIngestRequest, request: Request) -> BatchIngestResponse:
    batch_id = str(uuid.uuid4())
    vault_db.insert_batch(batch_id=batch_id, total_files=len(body.file_ids))
    traceparent = request.headers.get("traceparent", "")
    queued: list[str] = []
    skipped: list[BatchIngestSkippedItem] = []

    for file_id in body.file_ids:
        row = vault_db.get_file_by_id(file_id)
        if not row or row["index_status"] == "deleted":
            skipped.append(BatchIngestSkippedItem(file_id=file_id, reason="File not found"))
            continue
        if row.get("ingest_lock_job_id") or row["index_status"] == "pending":
            skipped.append(
                BatchIngestSkippedItem(file_id=file_id, reason="Ingest already in progress")
            )
            continue
        try:
            job_id = enqueue_vault_ingest(file_id, traceparent=traceparent)
        except ValueError as exc:
            skipped.append(BatchIngestSkippedItem(file_id=file_id, reason=str(exc)))
            continue
        vault_db.add_batch_file(batch_id, file_id, job_id)
        queued.append(file_id)

    vault_db.refresh_batch_counts(batch_id)
    return BatchIngestResponse(batch_id=batch_id, queued=queued, skipped=skipped)


@router.get("/files/{file_id}", response_model=VaultFile)
def get_file(file_id: str) -> VaultFile:
    try:
        return vault_store.get_file(file_id)
    except VaultStoreError as exc:
        _raise_store(exc)
        raise  # pragma: no cover


@router.get("/files/{file_id}/content")
def get_file_content(file_id: str) -> dict[str, str | int]:
    try:
        content = vault_store.read_file_content(file_id)
        file = vault_store.get_file(file_id)
    except VaultStoreError as exc:
        _raise_store(exc)
        raise  # pragma: no cover
    return {
        "id": file.id,
        "relative_path": file.relative_path,
        "content": content,
        "size": len(content.encode("utf-8")),
    }


@router.put("/files/{file_id}/content", response_model=UploadResponse)
def put_file_content(
    file_id: str, body: SaveContentRequest, request: Request
) -> UploadResponse:
    try:
        result = vault_store.save_file_content(file_id, body.content)
    except VaultStoreError as exc:
        _raise_store(exc)
        raise  # pragma: no cover
    job_id: str | None = None
    if result.auto_ingest:
        traceparent = request.headers.get("traceparent", "")
        job_id = _enqueue_ingest_or_409(file_id, traceparent=traceparent)
    return UploadResponse(file=vault_store.get_file(file_id), ingest_job_id=job_id)


@router.delete("/files/batch", response_model=BatchDeleteResult)
def delete_files_batch(body: BatchDeleteRequest) -> BatchDeleteResult:
    results: list[BatchDeleteItem] = []
    for file_id in body.file_ids:
        try:
            vault_store.delete_file(file_id)
            results.append(BatchDeleteItem(file_id=file_id, ok=True))
        except VaultStoreError as exc:
            results.append(
                BatchDeleteItem(file_id=file_id, ok=False, error=str(exc))
            )
    return BatchDeleteResult(results=results)


@router.post("/files/{file_id}/ingest", response_model=ReindexResponse)
def ingest_file(file_id: str, request: Request) -> ReindexResponse:
    try:
        file = vault_store.get_file(file_id)
    except VaultStoreError as exc:
        _raise_store(exc)
        raise  # pragma: no cover
    if file.ingest_locked:
        raise HTTPException(status_code=409, detail="File is locked while ingest is running")
    traceparent = request.headers.get("traceparent", "")
    job_id = _enqueue_ingest_or_409(file_id, traceparent=traceparent)
    return ReindexResponse(
        file_id=file_id,
        relative_path=file.relative_path,
        ingest_job_id=job_id,
    )


@router.post("/files/{file_id}/reindex", response_model=ReindexResponse)
def reindex_file(file_id: str, request: Request) -> ReindexResponse:
    """Alias for manual ingest (Phase 1.7)."""
    return ingest_file(file_id, request)


@router.post("/files/{file_id}/clear-index", response_model=ClearIndexResponse)
def clear_file_index(file_id: str) -> ClearIndexResponse:
    try:
        file = vault_store.clear_file_index(file_id)
    except VaultStoreError as exc:
        _raise_store(exc)
        raise  # pragma: no cover
    return ClearIndexResponse(
        file_id=file.id,
        relative_path=file.relative_path,
        index_status=file.index_status,
    )


@router.post("/sync", response_model=SyncReport)
def post_sync() -> SyncReport:
    return sync_vault()


@router.post("/migrate-v16", response_model=MigrateV16Response)
def migrate_vault_v16(
    purge_mode: str = Query(default="vault", pattern="^(vault|all)$"),
    dry_run: bool = Query(default=False),
) -> MigrateV16Response:
    if any_ingest_locked():
        raise HTTPException(
            status_code=409,
            detail="Vault migration blocked: one or more files have ingest in progress",
        )
    report = run_full_vault_migration(
        dry_run=dry_run,
        purge_mode=purge_mode,  # type: ignore[arg-type]
        skip_reindex=True,
    )
    return MigrateV16Response(
        total_files=report.total_files,
        job_ids=[
            MigrateV16JobEntry(
                file_id=entry.file_id,
                relative_path=entry.relative_path,
                ingest_job_id=entry.ingest_job_id,
            )
            for entry in report.job_ids
        ],
        dry_run=dry_run,
        neo4j_stats=report.neo4j_stats,
        redis_keys_deleted=report.redis_keys_deleted,
    )


@router.get("/batches/{batch_id}", response_model=BatchStatusResponse)
def get_batch(batch_id: str) -> BatchStatusResponse:
    batch = vault_db.get_batch(batch_id)
    if not batch:
        raise HTTPException(status_code=404, detail="Batch not found")
    vault_db.refresh_batch_counts(batch_id)
    batch = vault_db.get_batch(batch_id)
    assert batch is not None
    file_rows = vault_db.list_batch_files(batch_id)
    files: list[BatchUploadFileResult] = []
    completed = 0
    failed = 0
    for fr in file_rows:
        status = fr["status"]
        file_row = vault_db.get_file_by_id(fr["file_id"])
        if file_row:
            if file_row["index_status"] == "indexed":
                status = "completed"
                completed += 1
                vault_db.update_batch_file_status(batch_id, fr["file_id"], status="completed")
            elif file_row["index_status"] == "error":
                status = "failed"
                failed += 1
                vault_db.update_batch_file_status(batch_id, fr["file_id"], status="failed")
            elif file_row["index_status"] == "pending":
                status = "pending"
        files.append(
            BatchUploadFileResult(
                file_id=fr["file_id"],
                filename=file_row["filename"] if file_row else fr["file_id"],
                job_id=fr.get("job_id"),
                status=status,
            )
        )
    vault_db.refresh_batch_counts(batch_id)
    batch = vault_db.get_batch(batch_id)
    assert batch is not None
    return BatchStatusResponse(
        id=batch["id"],
        created_at=batch["created_at"],
        total_files=batch["total_files"],
        completed_files=batch["completed_files"],
        failed_files=batch["failed_files"],
        files=files,
    )

File: src/lib/api/vault.ts
/** RAG Content Vault API client — `/api/v1/rag/vault/*` */

export type VaultIndexStatus =
  | "not_indexed"
  | "pending"
  | "indexed"
  | "modified"
  | "error"
  | "deleted";

export interface VaultFolder {
  id: string;
  name: string;
  slug: string;
  created_at: string;
  relative_path: string;
  file_count: number;
}

export interface FolderRenamePreviewFile {
  old_relative_path: string;
  new_relative_path: string;
  index_status: VaultIndexStatus;
}

export interface FolderRenamePreview {
  folder_id: string;
  old_name: string;
  new_name: string;
  old_slug: string;
  new_slug: string;
  slug_unchanged: boolean;
  can_rename: boolean;
  block_reason?: string | null;
  total_files: number;
  neo4j_knowledge_count: number;
  preview_files: FolderRenamePreviewFile[];
  has_more_files: boolean;
}

export interface VaultFile {
  id: string;
  folder_id: string;
  filename: string;
  relative_path: string;
  source: string;
  created_at: string;
  updated_at: string;
  size_bytes: number;
  mime_ext: string;
  mutable: boolean;
  index_status: VaultIndexStatus;
  chunk_count: number;
  last_ingest_job_id?: string | null;
  last_ingest_at?: string | null;
  ingest_locked: boolean;
  error_message?: string | null;
  content_preview?: string | null;
}

export interface PaginatedFilesResponse {
  files: VaultFile[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface SyncReport {
  files_scanned: number;
  drift_added: number;
  drift_modified: number;
  drift_removed: number;
  last_sync_at?: string | null;
}

export interface UploadResponse {
  file: VaultFile;
  ingest_job_id?: string | null;
  replaced?: boolean;
}

export interface BatchUploadFileResult {
  file_id: string;
  filename: string;
  job_id?: string | null;
  status: string;
  error_message?: string | null;
}

export interface BatchUploadResponse {
  batch_id: string;
  files: BatchUploadFileResult[];
}

export interface BatchDeleteResult {
  results: { file_id: string; ok: boolean; error?: string | null }[];
}

const API_BASE = "/api/v1/rag/vault";

async function handleResponse<T>(res: Response): Promise<T> {
  if (!res.ok) {
    const body = await res.text();
    throw new Error(body || res.statusText);
  }
  if (res.status === 204) {
    return undefined as T;
  }
  return res.json() as Promise<T>;
}

export async function listFolders(): Promise<VaultFolder[]> {
  const res = await fetch(`${API_BASE}/folders`);
  const data = await handleResponse<{ folders: VaultFolder[] }>(res);
  return data.folders;
}

export async function createFolder(name: string): Promise<VaultFolder> {
  const res = await fetch(`${API_BASE}/folders`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name }),
  });
  return handleResponse<VaultFolder>(res);
}

export async function previewFolderRename(
  folderId: string,
  name: string
): Promise<FolderRenamePreview> {
  const qs = new URLSearchParams({ name });
  const res = await fetch(`${API_BASE}/folders/${folderId}/rename-preview?${qs}`);
  return handleResponse<FolderRenamePreview>(res);
}

export async function renameFolder(folderId: string, name: string): Promise<VaultFolder> {
  const res = await fetch(`${API_BASE}/folders/${folderId}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name }),
  });
  return handleResponse<VaultFolder>(res);
}

export async function deleteFolder(folderId: string): Promise<void> {
  const res = await fetch(`${API_BASE}/folders/${folderId}`, { method: "DELETE" });
  await handleResponse<void>(res);
}

export async function listVaultFiles(params: {
  folder_id?: string;
  keyword?: string;
  index_status?: string;
  search_content?: boolean;
  page?: number;
  page_size?: number;
}): Promise<PaginatedFilesResponse> {
  const search = new URLSearchParams();
  if (params.folder_id) search.set("folder_id", params.folder_id);
  if (params.keyword) search.set("keyword", params.keyword);
  if (params.index_status) search.set("index_status", params.index_status);
  if (params.search_content) search.set("search_content", "true");
  if (params.page) search.set("page", String(params.page));
  if (params.page_size) search.set("page_size", String(params.page_size));
  const qs = search.toString();
  const res = await fetch(`${API_BASE}/files${qs ? `?${qs}` : ""}`);
  return handleResponse<PaginatedFilesResponse>(res);
}

export async function createVaultFile(body: {
  folder_id: string;
  filename: string;
  content?: string;
}): Promise<UploadResponse> {
  const res = await fetch(`${API_BASE}/files`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  return handleResponse<UploadResponse>(res);
}

export async function uploadVaultFile(
  folderId: string,
  file: File
): Promise<UploadResponse> {
  const form = new FormData();
  form.append("folder_id", folderId);
  form.append("file", file);
  const res = await fetch(`${API_BASE}/files/upload`, { method: "POST", body: form });
  return handleResponse<UploadResponse>(res);
}

export async function batchUploadVaultFiles(
  folderId: string,
  files: File[]
): Promise<BatchUploadResponse> {
  const form = new FormData();
  form.append("folder_id", folderId);
  for (const f of files) {
    form.append("files", f);
  }
  const res = await fetch(`${API_BASE}/files/batch-upload`, { method: "POST", body: form });
  return handleResponse<BatchUploadResponse>(res);
}

export async function getVaultFileContent(fileId: string): Promise<{
  id: string;
  relative_path: string;
  content: string;
  size: number;
}> {
  const res = await fetch(`${API_BASE}/files/${fileId}/content`);
  return handleResponse(res);
}

export async function getVaultFileByPath(relativePath: string): Promise<VaultFile | null> {
  const listed = await listVaultFiles({ keyword: relativePath.split("/").pop(), page_size: 20 });
  return listed.files.find((f) => f.relative_path === relativePath) ?? null;
}

export async function saveVaultFileContent(
  fileId: string,
  content: string
): Promise<UploadResponse> {
  const res = await fetch(`${API_BASE}/files/${fileId}/content`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ content }),
  });
  return handleResponse<UploadResponse>(res);
}

export async function batchDeleteVaultFiles(fileIds: string[]): Promise<BatchDeleteResult> {
  const res = await fetch(`${API_BASE}/files/batch`, {
    method: "DELETE",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ file_ids: fileIds }),
  });
  return handleResponse<BatchDeleteResult>(res);
}

export async function reindexVaultFile(fileId: string): Promise<{
  file_id: string;
  relative_path: string;
  ingest_job_id: string;
}> {
  const res = await fetch(`${API_BASE}/files/${fileId}/reindex`, { method: "POST" });
  return handleResponse(res);
}

export interface IngestPreviewItem {
  file_id: string;
  relative_path: string;
  estimated_tokens: number;
  ingestible: boolean;
  block_reason?: string | null;
}

export interface IngestPreviewResponse {
  items: IngestPreviewItem[];
  total_estimated_tokens: number;
  file_count: number;
}

export async function previewIngest(fileIds: string[]): Promise<IngestPreviewResponse> {
  const res = await fetch(`${API_BASE}/files/ingest-preview`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ file_ids: fileIds }),
  });
  return handleResponse(res);
}

export async function ingestVaultFile(fileId: string): Promise<{
  file_id: string;
  relative_path: string;
  ingest_job_id: string;
}> {
  const res = await fetch(`${API_BASE}/files/${fileId}/ingest`, { method: "POST" });
  return handleResponse(res);
}

export interface BatchIngestSkippedItem {
  file_id: string;
  reason: string;
}

export interface BatchIngestResponse {
  batch_id: string;
  queued: string[];
  skipped: BatchIngestSkippedItem[];
}

export async function batchIngestVaultFiles(fileIds: string[]): Promise<BatchIngestResponse> {
  const res = await fetch(`${API_BASE}/files/batch-ingest`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ file_ids: fileIds }),
  });
  return handleResponse(res);
}

export async function clearVaultIndex(fileId: string): Promise<{
  file_id: string;
  relative_path: string;
  index_status: VaultIndexStatus;
}> {
  const res = await fetch(`${API_BASE}/files/${fileId}/clear-index`, { method: "POST" });
  return handleResponse(res);
}

export interface MigrateV16JobEntry {
  file_id: string;
  relative_path: string;
  ingest_job_id: string;
}

export interface MigrateV16Response {
  total_files: number;
  job_ids: MigrateV16JobEntry[];
  dry_run?: boolean;
  neo4j_stats?: Record<string, number>;
  redis_keys_deleted?: number;
}

export async function migrateVaultV16(params?: {
  purge_mode?: "vault" | "all";
  dry_run?: boolean;
}): Promise<MigrateV16Response> {
  const search = new URLSearchParams();
  if (params?.purge_mode) search.set("purge_mode", params.purge_mode);
  if (params?.dry_run) search.set("dry_run", "true");
  const qs = search.toString();
  const res = await fetch(`${API_BASE}/migrate-v16${qs ? `?${qs}` : ""}`, { method: "POST" });
  return handleResponse<MigrateV16Response>(res);
}

export async function syncVault(): Promise<SyncReport> {
  const res = await fetch(`${API_BASE}/sync`, { method: "POST" });
  return handleResponse<SyncReport>(res);
}

export async function getBatchStatus(batchId: string) {
  const res = await fetch(`${API_BASE}/batches/${batchId}`);
  return handleResponse<{
    id: string;
    total_files: number;
    completed_files: number;
    failed_files: number;
    files: BatchUploadFileResult[];
  }>(res);
}

export function languageFromPath(path: string): string {
  if (path.endsWith(".md")) return "markdown";
  if (path.endsWith(".py")) return "python";
  if (path.endsWith(".sql")) return "sql";
  return "plaintext";
}
